src/driver/op.rs

```diff
@@ -183,6 +191,76 @@ where
     }
 }
 
+       
+impl<T> Future for Op<T, MultiCQEFuture>
+where
+    T: Unpin + 'static + Completable + Updateable,
+{
+    type Output = T::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        use std::mem;
+
+        let me = &mut *self;
+
+        CONTEXT.with(|runtime_context| {
+            runtime_context.with_driver_mut(|driver| {
+                let (lifecycle, completions) = driver
+                    .ops
+                    .get_mut(me.index)
+                    .expect("invalid internal state");
+
+                match mem::replace(lifecycle, Lifecycle::Submitted) {
+                    Lifecycle::Submitted => {
+                        *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                        Poll::Pending
+                    }
+                    Lifecycle::Waiting(waker) if !waker.will_wake(cx.waker()) => {
+                        *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                        Poll::Pending
+                    }
+                    Lifecycle::Waiting(waker) => {
+                        *lifecycle = Lifecycle::Waiting(waker);
+                        Poll::Pending
+                    }
+                    Lifecycle::Ignored(..) => unreachable!(),
+                    Lifecycle::Completed(cqe) => {
+                        // This is possible. We may have previously polled a CompletionList,
+                        // and the final CQE registered as Completed
+                        driver.ops.remove(me.index);
+                        me.index = usize::MAX;
+                        Poll::Ready(me.data.take().unwrap().complete(cqe))
+                    }
+                    Lifecycle::CompletionList(indices) => {
+                        let mut data = me.data.take().unwrap();
+
+                        // Traverse the CqeResult list, calling update on the Op on all Cqe's flagged `more`
+                        // If the final Cqe is present, return it in an Option
+                        let last = indices.into_list(completions).find(|cqe| {
```

Comment:
This find doesn't actually look at all the entries does it? It finishes as soon as the closure returns true, returning that cqe. So theoretically more than one entry could have no 'more' flag set. I don't quite know what to make of that. Just seems like things aren't completely tight. Just thinking out load here.

Reply:
NN o, but then if we're receiving more cqes without more flagged, there is a bug in the kernel